### ISSUE TITLE
Improve database query

### ISSUE THREAD

**Issue by rh-hemartin** [kubearchive/kubearchive#580]
Querying resources without namespace is slow and its not using the index we created. Using the production dataset we have we can reproduce this easily.

This is our current query.
```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY ((data->'metadata'->'creationTimestamp')::text::timestamptz, uuid) LIMIT 100;
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=494057.98..494069.64 rows=100 width=395) (actual time=18696.120..18705.782 rows=100 loops=1)
   ->  Gather Merge  (cost=494057.98..523553.14 rows=252798 width=395) (actual time=18681.244..18690.901 rows=100 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=493057.95..493373.95 rows=126399 width=395) (actual time=18657.673..18657.676 rows=74 loops=3)
               Sort Key: (ROW(((((data -> 'metadata'::text) -> 'creationTimestamp'::text))::text)::timestamp with time zone, uuid))
               Sort Method: top-N heapsort  Memory: 43kB
               Worker 0:  Sort Method: top-N heapsort  Memory: 42kB
               Worker 1:  Sort Method: top-N heapsort  Memory: 43kB
               ->  Parallel Bitmap Heap Scan on resource  (cost=4153.85..488227.07 rows=126399 width=395) (actual time=62.448..18551.861 rows=203719 loops=3)
                     Recheck Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
                     Rows Removed by Index Recheck: 1318857
                     Heap Blocks: exact=11655 lossy=88274
                     ->  Bitmap Index Scan on resource_kind_namespace_idx  (cost=0.00..4078.01 rows=303358 width=0) (actual time=75.179..75.179 rows=611157 loops=1)
                           Index Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
 Planning Time: 0.295 ms
 JIT:
   Functions: 13
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 2.171 ms (Deform 0.826 ms), Inlining 0.000 ms, Optimization 1.243 ms, Emission 23.204 ms, Total 26.619 ms
 Execution Time: 18707.027 ms
(21 rows)
```

Acceptance Criteria:

* Improve the time for this query.

**Comment by rh-hemartin** [comment]
Our database expert recommended us to do a few changes, regarding the query itself and the index we are using. Given that and other sources, in the following comments I will be doing one single change and then pasting results, so we can see the progression of the improvement.

**Comment by ggallen** [comment]
> Our database expert recommended us to do a few changes, regarding the query itself and the index we are using.

Can you detail the suggested changes?

**Comment by rh-hemartin** [comment]
* Use text to order rather than casting to `timestamptz`, so remove `::text::timestamptz` from the query. RFC3339 can be sorted as text given all timestamps are formatted the same, in our case with `Z` and everyone with the same precision. (https://www.rfc-editor.org/rfc/rfc3339#section-5.1)

```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY (data->'metadata'->'creationTimestamp', uuid) LIMIT 100;
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=492793.99..492805.65 rows=100 width=395) (actual time=18673.856..18682.562 rows=100 loops=1)
   ->  Gather Merge  (cost=492793.99..522289.15 rows=252798 width=395) (actual time=18659.876..18668.576 rows=100 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=491793.96..492109.96 rows=126399 width=395) (actual time=18641.498..18641.504 rows=81 loops=3)
               Sort Key: (ROW(((data -> 'metadata'::text) -> 'creationTimestamp'::text), uuid))
               Sort Method: top-N heapsort  Memory: 62kB
               Worker 0:  Sort Method: top-N heapsort  Memory: 58kB
               Worker 1:  Sort Method: top-N heapsort  Memory: 60kB
               ->  Parallel Bitmap Heap Scan on resource  (cost=4153.85..486963.08 rows=126399 width=395) (actual time=54.987..18455.828 rows=203719 loops=3)
                     Recheck Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
                     Rows Removed by Index Recheck: 1318857
                     Heap Blocks: exact=11572 lossy=88197
                     ->  Bitmap Index Scan on resource_kind_namespace_idx  (cost=0.00..4078.01 rows=303358 width=0) (actual time=63.069..63.069 rows=611157 loops=1)
                           Index Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
 Planning Time: 0.318 ms
 JIT:
   Functions: 13
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 1.878 ms (Deform 0.708 ms), Inlining 0.000 ms, Optimization 1.207 ms, Emission 21.807 ms, Total 24.893 ms
 Execution Time: 18683.698 ms
(21 rows)
```

**Comment by rh-hemartin** [comment]
* Use PostgreSQL "to text" json operator (https://www.postgresql.org/docs/current/functions-json.html), so use `->>` instead of `->`.

```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY (data->'metadata'->>'creationTimestamp', uuid) LIMIT 100;
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=492793.99..492805.65 rows=100 width=395) (actual time=18579.232..18589.868 rows=100 loops=1)
   ->  Gather Merge  (cost=492793.99..522289.15 rows=252798 width=395) (actual time=18575.488..18586.118 rows=100 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=491793.96..492109.96 rows=126399 width=395) (actual time=18561.454..18561.458 rows=75 loops=3)
               Sort Key: (ROW(((data -> 'metadata'::text) ->> 'creationTimestamp'::text), uuid))
               Sort Method: top-N heapsort  Memory: 62kB
               Worker 0:  Sort Method: top-N heapsort  Memory: 58kB
               Worker 1:  Sort Method: top-N heapsort  Memory: 60kB
               ->  Parallel Bitmap Heap Scan on resource  (cost=4153.85..486963.08 rows=126399 width=395) (actual time=51.745..18401.928 rows=203719 loops=3)
                     Recheck Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
                     Rows Removed by Index Recheck: 1318857
                     Heap Blocks: exact=11482 lossy=88300
                     ->  Bitmap Index Scan on resource_kind_namespace_idx  (cost=0.00..4078.01 rows=303358 width=0) (actual time=55.806..55.806 rows=611157 loops=1)
                           Index Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
 Planning Time: 0.096 ms
 JIT:
   Functions: 13
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 1.128 ms (Deform 0.407 ms), Inlining 0.000 ms, Optimization 0.731 ms, Emission 12.150 ms, Total 14.009 ms
 Execution Time: 18590.246 ms
(21 rows)
```

**Comment by rh-hemartin** [comment]
* Remove the composite in `ORDER BY`, so `ORDER BY timestamp, uuid` instead of `ORDER BY (timestamp, uuid)` (https://www.postgresql.org/docs/17/rowtypes.html)
```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY data->'metadata'->>'creationTimestamp', uuid LIMIT 100;
                                                                            QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=492793.99..492805.65 rows=100 width=395) (actual time=18633.909..18643.653 rows=100 loops=1)
   ->  Gather Merge  (cost=492793.99..522289.15 rows=252798 width=395) (actual time=18620.387..18630.125 rows=100 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=491793.96..492109.96 rows=126399 width=395) (actual time=18594.529..18594.533 rows=78 loops=3)
               Sort Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text)), uuid
               Sort Method: top-N heapsort  Memory: 44kB
               Worker 0:  Sort Method: top-N heapsort  Memory: 42kB
               Worker 1:  Sort Method: top-N heapsort  Memory: 42kB
               ->  Parallel Bitmap Heap Scan on resource  (cost=4153.85..486963.08 rows=126399 width=395) (actual time=59.519..18479.174 rows=203719 loops=3)
                     Recheck Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
                     Rows Removed by Index Recheck: 1318857
                     Heap Blocks: exact=11619 lossy=89137
                     ->  Bitmap Index Scan on resource_kind_namespace_idx  (cost=0.00..4078.01 rows=303358 width=0) (actual time=75.120..75.121 rows=611157 loops=1)
                           Index Cond: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
 Planning Time: 0.551 ms
 JIT:
   Functions: 13
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 1.928 ms (Deform 0.788 ms), Inlining 0.000 ms, Optimization 1.321 ms, Emission 21.167 ms, Total 24.416 ms
 Execution Time: 18644.870 ms
(21 rows)
```

**Comment by rh-hemartin** [comment]
* Create the index: `CREATE INDEX idx_creation_timestamp_btree ON public.resource USING btree ((data->'metadata'->>'creationTimestamp'));`

```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY data->'metadata'->>'creationTimestamp', uuid LIMIT 100;
                                                                            QUERY PLAN                                                                             
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=7.43..710.83 rows=100 width=395) (actual time=14.661..37.044 rows=100 loops=1)
   ->  Incremental Sort  (cost=7.43..2133826.78 rows=303358 width=395) (actual time=14.659..37.036 rows=100 loops=1)
         Sort Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text)), uuid
         Presorted Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text))
         Full-sort Groups: 4  Sort Method: quicksort  Average Memory: 29kB  Peak Memory: 29kB
         ->  Index Scan using idx_creation_timestamp_btree on resource  (cost=0.43..2120175.67 rows=303358 width=395) (actual time=1.271..36.878 rows=101 loops=1)
               Filter: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
               Rows Removed by Filter: 1115
 Planning Time: 0.295 ms
 Execution Time: 37.088 ms
(10 rows)
```

**Comment by rh-hemartin** [comment]
TLDR: applying the follow changes to the query it improves 3 orders of magnitude: from tens of seconds, to tens of miliseconds.

* Use text to order rather than casting to `timestamptz`, so remove `::text::timestamptz` from the query. RFC3339 can be sorted as text given all timestamps are formatted the same, in our case with `Z` and everyone with the same precision. (https://www.rfc-editor.org/rfc/rfc3339#section-5.1)
* Use PostgreSQL "to text" json operator (https://www.postgresql.org/docs/current/functions-json.html), so use `->>` instead of `->`.
* Remove the composite in `ORDER BY`, so `ORDER BY timestamp, uuid` instead of `ORDER BY (timestamp, uuid)` (https://www.postgresql.org/docs/17/rowtypes.html)
* Create the index: `CREATE INDEX idx_creation_timestamp_btree ON public.resource USING btree ((data->'metadata'->>'creationTimestamp'));`


```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY data->'metadata'->>'creationTimestamp', uuid LIMIT 100;
                                                                            QUERY PLAN                                                                             
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=7.43..710.83 rows=100 width=395) (actual time=14.661..37.044 rows=100 loops=1)
   ->  Incremental Sort  (cost=7.43..2133826.78 rows=303358 width=395) (actual time=14.659..37.036 rows=100 loops=1)
         Sort Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text)), uuid
         Presorted Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text))
         Full-sort Groups: 4  Sort Method: quicksort  Average Memory: 29kB  Peak Memory: 29kB
         ->  Index Scan using idx_creation_timestamp_btree on resource  (cost=0.43..2120175.67 rows=303358 width=395) (actual time=1.271..36.878 rows=101 loops=1)
               Filter: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
               Rows Removed by Filter: 1115
 Planning Time: 0.295 ms
 Execution Time: 37.088 ms
(10 rows)
```

**Comment by rh-hemartin** [comment]
I found that we need to sort the table DESC by timestamp, so we get the newer ones at the top. I tested with both the index in DESC order already and the index in ASC order (the default order), there are the results:

Index DESC:

```
kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY data->'metadata'->>'creationTimestamp', uuid LIMIT 100;
                                                                              QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=7.43..710.83 rows=100 width=395) (actual time=11.967..37.944 rows=100 loops=1)
   ->  Incremental Sort  (cost=7.43..2133826.78 rows=303358 width=395) (actual time=11.966..37.935 rows=100 loops=1)
         Sort Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text)), uuid
         Presorted Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text))
         Full-sort Groups: 4  Sort Method: quicksort  Average Memory: 29kB  Peak Memory: 29kB
         ->  Index Scan Backward using idx_creation_timestamp on resource  (cost=0.43..2120175.67 rows=303358 width=395) (actual time=0.562..37.778 rows=101 loops=1)
               Filter: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
               Rows Removed by Filter: 1115
 Planning Time: 0.125 ms
 Execution Time: 37.975 ms
(10 rows)

kubearchive=# EXPLAIN (ANALYZE) SELECT data->'metadata'->'creationTimestamp', uuid, data FROM resource WHERE kind='PipelineRun' AND api_version='tekton.dev/v1beta1' ORDER BY data->'metadata'->>'creationTimestamp' DESC, uuid LIMIT 100;
                                                                            QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=7.43..710.83 rows=100 width=395) (actual time=1714.394..1720.583 rows=100 loops=1)
   ->  Incremental Sort  (cost=7.43..2133826.78 rows=303358 width=395) (actual time=1714.393..1720.576 rows=100 loops=1)
         Sort Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text)) DESC, uuid
         Presorted Key: (((data -> 'metadata'::text) ->> 'creationTimestamp'::text))
         Full-sort Groups: 4  Sort Method: quicksort  Average Memory: 29kB  Peak Memory: 29kB
         ->  Index Scan using idx_creation_timestamp on resource  (cost=0.43..2120175.67 rows=303358 width=395) (actual time=1709.851..1720.427 rows=101 loops=1)
               Filter: (((kind)::text = 'PipelineRun'::text) AND ((api_version)::text = 'tekton.dev/v1beta1'::text))
               Rows Removed by Filter: 3841646
 Planning Time: 0.096 ms
 Execution Time: 1720.605 ms
(10 rows)
```

Even if the index is in the same direction (DESC), the query takes a little bit longer. Probably because how the table was built. In the DESC query we get `Rows Removed by Filter: 3841646`, while in the ASC query we get `ows Removed by Filter: 1115`.

Note: The results are the same for both ASC and DESC queries regardless of the direction of the index.

**Comment by rh-hemartin** [comment]
I tried the DESC ordering while doing a continue type of query and there are problems because of the sorting of uuids:

* If we keep the combined condition (inverted because of DESC) `(timestamp, uuid) < (value, value)` the first and second page of the pagination get some overlapped results due to the sorting of uuids.
* If we uncoupled the condition to be (timestamp < value AND uuid > value) the second page of the pagination loses some entries.

Tests where done getting two consecutive pages of 100 and one page of 200 and comparing their entries on the production database using the `crt-nshift-lightspeed-tenant` namespace. The queries contained the changes listed on the TLDR (`->>`, order by without composite, index, no casting) and then DESC and DESC + decoupled AND condition.
